src/iam/iam-base.ts

```diff
@@ -245,38 +255,97 @@ export class IAMBase {
         });
       }
       this._signer = new providers.Web3Provider(metamaskProvider).getSigner();
+
+      metamaskProvider.on("networkChanged", () => {
+        this.closeConnection();
+      });
+      metamaskProvider.on("accountsChanged", () => {
+        this.closeConnection();
+      });
+
+      this._providerType = walletProvider;
       return;
     }
-    this._transactionOverrides = {
-      gasLimit: hexlify(4900000),
-      gasPrice: hexlify(0.1)
-    };
+    if (
+      walletProvider === WalletProvider.EwKeyManager ||
+      walletProvider === WalletProvider.WalletConnect
+    ) {
+      this._transactionOverrides = {
+        gasLimit: hexlify(4900000),
+        gasPrice: hexlify(0.1)
+      };
 
-    const showQRCode = !(walletProvider === WalletProvider.EwKeyManager);
-
-    this._walletConnectProvider = new WalletConnectProvider({
-      rpc: {
-        [this._connectionOptions.chainId]: this._connectionOptions.rpcUrl
-      },
-      infuraId: this._connectionOptions.infuraId,
-      bridge: this._connectionOptions.bridgeUrl,
-      qrcode: showQRCode
-    });
-
-    if (walletProvider === WalletProvider.EwKeyManager) {
-      this._walletConnectProvider.wc.on("display_uri", (err, payload) => {
-        // uri is expected to be WalletConnect Standard URI https://eips.ethereum.org/EIPS/eip-1328
-        const wcUri = payload.params[0];
-
-        const encoded = encodeURIComponent(wcUri);
-        const hasQueryString = this._ewKeyManagerUrl.includes("?");
-        const url = `${this._ewKeyManagerUrl}${hasQueryString ? "&" : "?"}uri=${encoded}`;
-        window.open(url, "_blank");
+      const showQRCode = !(walletProvider === WalletProvider.EwKeyManager);
+
+      this._walletConnectProvider = new WalletConnectProvider({
+        rpc: {
+          [this._connectionOptions.chainId]: this._connectionOptions.rpcUrl
+        },
+        infuraId: this._connectionOptions.infuraId,
+        bridge: this._connectionOptions.bridgeUrl,
+        qrcode: showQRCode
       });
+
+      if (walletProvider === WalletProvider.EwKeyManager) {
+        this._walletConnectProvider.wc.on("display_uri", (err, payload) => {
+          // uri is expected to be WalletConnect Standard URI https://eips.ethereum.org/EIPS/eip-1328
+          const wcUri = payload.params[0];
+
+          const encoded = encodeURIComponent(wcUri);
+          const hasQueryString = this._ewKeyManagerUrl.includes("?");
+          const url = `${this._ewKeyManagerUrl}${hasQueryString ? "&" : "?"}uri=${encoded}`;
+          window.open(url, "_blank");
+        });
+      }
+
+      await this._walletConnectProvider.enable();
+
+      this._walletConnectProvider.wc.on("session_update", () => {
+        this.closeConnection();
+      });
+
+      this._walletConnectProvider.wc.on("disconnect", () => {
+        this.closeConnection();
+      });
```

Comment:
Just curious, are these eventListeners set after `walletConnectProvider.enable` for a reason? In a situation where disconnect is emitted immediately after enable, would we want to be listening for that event?

Reply:
After second though the listeners don't need to be set after the enabling the `walletConnectProvider` , in this spring we could add a `on` method to be able to add listeners for certain events.